scripts/o3de/o3de/download.py

```diff
@@ -149,38 +151,108 @@ def download_o3de_object(object_name: str, default_folder_name: str, dest_path:
 def download_gem(gem_name: str,
                  dest_path: str or pathlib.Path,
                  skip_auto_register: bool,
+                 force_overwrite: bool,
                  download_progress_callback = None) -> int:
-    return download_o3de_object(gem_name, 'gems', dest_path, 'gem', 'gem_name', skip_auto_register, download_progress_callback)
+    return download_o3de_object(gem_name, 'gems', dest_path, 'gem', 'gem_name', skip_auto_register, force_overwrite, download_progress_callback)
 
 
 def download_template(template_name: str,
                       dest_path: str or pathlib.Path,
                       skip_auto_register: bool,
+                      force_overwrite: bool,
                       download_progress_callback = None) -> int:
-    return download_o3de_object(template_name, 'templates', dest_path, 'template', 'template_name', skip_auto_register, download_progress_callback)
+    return download_o3de_object(template_name, 'templates', dest_path, 'template', 'template_name', skip_auto_register, force_overwrite, download_progress_callback)
 
 
 
 def download_restricted(restricted_name: str,
                         dest_path: str or pathlib.Path,
                         skip_auto_register: bool,
+                        force_overwrite: bool,
                         download_progress_callback = None) -> int:
-    return download_o3de_object(restricted_name, 'restricted', dest_path, 'restricted', 'restricted_name', skip_auto_register, download_progress_callback)
+    return download_o3de_object(restricted_name, 'restricted', dest_path, 'restricted', 'restricted_name', skip_auto_register, force_overwrite, download_progress_callback)
+
+def is_o3de_object_update_available(object_name: str, downloadable_kwarg_key, current_date: datetime) -> bool:
+    downloadable_object_data = get_downloadable(**{downloadable_kwarg_key : object_name})
+    if not downloadable_object_data:
+        logger.error(f'Downloadable o3de object {object_name} not found.')
+        return False
+
+    repo_copy_updated_string = 0
+    try:
+        repo_copy_updated_string = downloadable_object_data['last_updated']
+    except KeyError:
+        logger.warn(f'last_updated field not found for {object_name}.')
+        return False
+
+    try:
+        repo_copy_updated_date = datetime.fromisoformat(repo_copy_updated_string)
+    except ValueError:
+        logger.error(f'last_updated field in incorrect format for {object_name}.')
+        return False
+
+    return repo_copy_updated_date > current_date
 
+def is_o3de_engine_update_available(engine_name: str, local_last_updated: str):
+    try:
+        current_updated_time = datetime.fromisoformat(local_last_updated)
+    except ValueError:
+        logger.warn(f'last_updated field has incorrect format for engine {engine_name}.')
+        # Possible that an earlier version did not have this field so still want to check against cached downloadable version
+        current_updated_time = datetime.min
```

Comment:
The querying of the time value from `local_last_updated` string can be done in the `is_o3de_object_update_available`

The only difference between the each implementation below is the o3de_object name.
The o3de object name itself is available via the `downloadable_kwarg_key` parameter in the `is_o3de_object_update_available` method

Reply:
+1

Reply:
Yes, good catch. Fixing this duplication also showed that the error emitted from both the local copy and the repo copy of this conversion were identical with no way to tell them apart.